### ISSUE TITLE
Touch input?

### ISSUE THREAD

**Issue by Jay2645** [segross/UnrealImGui#20]
It seems like ImGui is consuming all touch inputs on mobile. Is there a way to explicitly enable/disable touch input like what we have for keyboard/gamepad input?

**Comment by segross** [comment]
Right now I don’t plan any custom solution. I can only test it with touch emulation but if there is a clear vision how it should work, I may try to add it. Alternatively I may just put more effort in making it easier to extend or even replace the whole input widget. But I would hope that even now it should be easy enough to modify SImGuiWidget.

Coming to your problem, I’m not sure whether I fully understand it. Are you talking about issues with clicking ImGui or occluding Slate buttons? SImGuiWidget does not handle touch events, so are you talking about your implementation?

**Comment by Jay2645** [comment]
By default, touch is treated as a mouse move event all the time. If you tapped on something, rather than it being simulated as a mouse move + mouse click, it was just being simulated as a mouse move. I had a feeling it probably wouldn't be something that would get implemented, so I wound up making a custom InputHandler which fixed it. 

Another issue I was having was that by default ImGui would "eat" each touch event and it wouldn't get sent any further, even if no ImGui controls were running. For example, if I was using my finger to turn the camera, I would drag my finger across the screen and nothing would happen if the ImGui plugin was enabled. I fixed it by enabling mouse sharing all the time and then disabling it when a menu was open, which isn't ideal but *cie la vie*.

**Comment by segross** [comment]
That sounds worse, if it breaks a general flow. I'll see if I can use touch emulation to reproduce it.

Do you keep input mode enabled all the time? With input mode disabled, input events should not be processed nor consumed. Also, it sounds that you want to share input between ImGui and game, so I would think it makes sense to keep this option enabled. Hard to say before reproducing it but it feels that you may have a problem with occlusion. If touch event moves mouse above ImGui controls, then it will enable widget collisions until something else moves the cursor out. When hit tests are enabled then widget occludes other Slate controls, what means that they won't get mouse and in this case touch events. 

On a side note, may I ask whether you use ImGui for debugging or in-game menus?

**Comment by Jay2645** [comment]
I only enabled input mode when I needed to (i.e. when a menu was open). ImGui is just used for debug menus, so it's not anything critical.

I'm not sure how well you'd be able to reproduce the issue since the issue only appears on device, not within the editor. Even with touch emulation, I believe mouse events were still happening (although I haven't actually investigated why that is). I do also know that `SImGuiWidget` doesn't override or listen to Slate's touch events at all; part of the fix I had was implementing them. My fix was pretty hacky, though, which is why I haven't submitted a pull request.

**Comment by segross** [comment]
Yes, it may be hard to reproduce as it seems to always send the mouse and emulated touch events. Hard to work without more details but I suspect there might be two problems. One I already mentioned and it would be a mouse cursor left hovering ImGui control and in result keeping the widget hit-test visible. Another, potential and connected problem would be a behaviour in the widget code that re-takes focus when it is lost while in input mode (in current widget code line 434). It would be better if it was not needed but without this it is possible to lose focus after opening and closing console (in standalone, not in the editor), what I really do not like. With the exclusive input mode, it seems to work as expected but in other situations it may be problematic. That is why, when mouse is shared I disable it when possible. However, when cursor is hovering ImGui control it behaves like the default input, so it would be interesting to check whether disabling this behaviour helps.

I’ll keep this open in case somebody has similar problem and or more details.

One related question. Do you think that it would be easier to customize after splitting the widget for rendering and input widgets and allowing to replace the input one? That would be instead or alongside the input handler, probably in similar fashion to AHUD actors.

**Comment by Jay2645** [comment]
Yeah, I've noticed the issue with window focus myself, but I've just been dealing with it (in practice, you just need to double-tap the window to get an input in). I'll keep your note in mind when looking at that problem later, though!

I agree that the best approach is probably your idea of splitting the widget between input and rendering. Input would be exposed with `IMGUI_API` and accessible from other modules, so users could potentially extend it and modify input modes without needing to modify your code directly. That'll help with maintaining the plugin, since I wouldn't need to merge any changes I make to touch input with any future changes or bugfixes you'd make. My changes would live in their own module and would just override Unreal ImGui code rather than being direct changes to the plugin's code (since right now the widget isn't exposed to other modules in any way).

**Comment by segross** [comment]
I have pushed a change adding support for touch input. Events are registered and then used to mock mouse move and button-down. ImGui gets one extra frame of processing after touch ending to make sure that it can process mouse-up. This is WIP and for now it doesn’t share input - I will add it, if I find a moment.
